File: baseLib/src/main/java/me/ycdev/android/lib/common/async/TaskScheduler.java
package me.ycdev.android.lib.common.async;

import android.annotation.SuppressLint;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.os.SystemClock;

import androidx.annotation.IntDef;
import androidx.annotation.MainThread;
import androidx.annotation.NonNull;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.concurrent.atomic.AtomicInteger;

import androidx.annotation.VisibleForTesting;
import me.ycdev.android.lib.common.utils.DateTimeUtils;
import me.ycdev.android.lib.common.utils.Preconditions;
import timber.log.Timber;

import static me.ycdev.android.lib.common.async.TaskScheduler.SchedulePolicy.IGNORE;
import static me.ycdev.android.lib.common.async.TaskScheduler.SchedulePolicy.NO_CHECK;
import static me.ycdev.android.lib.common.async.TaskScheduler.SchedulePolicy.REPLACE;
import static me.ycdev.android.lib.common.utils.ThreadUtils.isMainThread;

@SuppressWarnings({"unused", "WeakerAccess"})
public class TaskScheduler {
    private static final String TAG = "TaskScheduler";

    @Retention(RetentionPolicy.SOURCE)
    @IntDef({NO_CHECK, IGNORE, REPLACE})
    @interface SchedulePolicy {
        int NO_CHECK = 1;
        int IGNORE = 2;
        int REPLACE = 3;
    }

    private static final int MSG_ADD_TASK = 1;
    private static final int MSG_REMOVE_TASK = 2;
    private static final int MSG_CHECK_TASKS = 3;
    private static final int MSG_CLEAR_TASKS = 4;

    @VisibleForTesting
    static final long DEFAULT_CHECK_INTERVAL = 10_000; // 10 seconds

    private static AtomicInteger sTaskSchedulerId = new AtomicInteger(1);

    private ITaskExecutor mTaskExecutor;
    private String mOwnerTag;
    private long mCheckInterval = DEFAULT_CHECK_INTERVAL;
    private boolean mLogEnabled = false;

    private Handler mMainHandler = new MainHandler();
    private ArrayList<TaskInfo> mTasks = new ArrayList<>();

    // for test only
    @VisibleForTesting int mCheckCount;

    public TaskScheduler(@NonNull ITaskExecutor executor, @NonNull String ownerTag) {
        Preconditions.checkNotNull(executor);
        Preconditions.checkNotNull(ownerTag);
        mTaskExecutor = executor;
        mOwnerTag = sTaskSchedulerId.getAndIncrement() + "-" + ownerTag;
    }

    public void setCheckInterval(long interval) {
        if (interval < 1000) {
            throw new IllegalArgumentException("Interval less than 1 second is not allowed.");
        }
        mCheckInterval = interval;
    }

    public void enableDebugLogs(boolean enable) {
        mLogEnabled = enable;
    }

    private static String schedulePolicyToString(@SchedulePolicy int policy) {
        switch (policy) {
            case NO_CHECK: return "NO_CHECK";
            case IGNORE: return "IGNORE";
            case REPLACE: return "REPLACE";
            default: throw new RuntimeException("Unknown policy: " + policy);
        }
    }

    private static void checkSchedulePolicy(@SchedulePolicy int policy) {
        switch (policy) {
            case NO_CHECK:
            case IGNORE:
            case REPLACE:
                return;
            default: throw new RuntimeException("Unknown policy: " + policy);
        }
    }

    public void scheduleAt(@NonNull Runnable task, long delayedMs) {
        scheduleAt(task, delayedMs, NO_CHECK);
    }

    public void scheduleAt(@NonNull Runnable task, long delayedMs, @SchedulePolicy int policy) {
        checkSchedulePolicy(policy);
        TaskInfo taskInfo = new TaskInfo(task, delayedMs);
        if (mLogEnabled) {
            Timber.tag(TAG).d("[%s] schedule one-off task: %s, policy: %s",
                    mOwnerTag, taskInfo, schedulePolicyToString(policy));
        }
        scheduleTask(taskInfo, policy);
    }

    public void schedulePeriod(@NonNull Runnable task, long delayedMs, long periodMs) {
        schedulePeriod(task, delayedMs, periodMs, NO_CHECK);
    }

    public void schedulePeriod(@NonNull Runnable task, long delayedMs, long periodMs,
            @SchedulePolicy int policy) {
        checkSchedulePolicy(policy);
        TaskInfo taskInfo = new TaskInfo(task, delayedMs, periodMs);
        if (mLogEnabled) {
            Timber.tag(TAG).d("[%s] schedule period task: %s, policy: %s",
                    mOwnerTag, taskInfo, schedulePolicyToString(policy));
        }
        scheduleTask(taskInfo, policy);
    }

    private void scheduleTask(TaskInfo taskInfo, @SchedulePolicy int policy) {
        if (isMainThread()) {
            addTask(taskInfo, policy);
        } else {
            mMainHandler.obtainMessage(MSG_ADD_TASK, policy, 0, taskInfo).sendToTarget();
        }
    }

    public void cancel(@NonNull Runnable task) {
        if (mLogEnabled) {
            Timber.tag(TAG).d("[%s] cancel task: %s", mOwnerTag, task);
        }
        if (isMainThread()) {
            removeTask(task);
        } else {
            mMainHandler.obtainMessage(MSG_REMOVE_TASK, task).sendToTarget();
        }
    }

    public void clear() {
        if (mLogEnabled) {
            Timber.tag(TAG).d("[%s] clear tasks", mOwnerTag);
        }
        if (isMainThread()) {
            clearTasks();
        } else {
            mMainHandler.sendEmptyMessage(MSG_CLEAR_TASKS);
        }
    }

    public void trigger() {
        if (mLogEnabled) {
            Timber.tag(TAG).d("[%s] trigger checking", mOwnerTag);
        }
        if (isMainThread()) {
            checkTasks();
        } else {
            mMainHandler.sendEmptyMessage(MSG_CHECK_TASKS);
        }
    }

    @MainThread
    private void addTask(TaskInfo task, @SchedulePolicy int policy) {
        boolean taskAdded = false;
        if (policy == NO_CHECK) {
            mTasks.add(task);
            taskAdded = true;
        } else {
            int index = findTaskIndex(task.task);
            if (index == -1) {
                mTasks.add(task);
                taskAdded = true;
            } else {
                if (mLogEnabled) {
                    Timber.tag(TAG).d("[%s] duplicate task found when add %s", mOwnerTag, task);
                }
                if (policy == REPLACE) {
                    mTasks.set(index, task);
                    taskAdded = true;
                } //else: nothing to do for ignore
            }
        }

        if (taskAdded) {
            scheduleCheckTask(task.delay);
            if (mLogEnabled) {
                Timber.tag(TAG).d("[%s] addTask: %s, policy: %s",
                        mOwnerTag, task, schedulePolicyToString(policy));
            }
        }
    }

    @MainThread
    private int findTaskIndex(@NonNull Runnable task) {
        for (int i = 0; i < mTasks.size(); i++) {
            TaskInfo info = mTasks.get(i);
            if (info.task.equals(task)) {
                return i;
            }
        }
        return -1;
    }

    @MainThread
    private void removeTask(@NonNull Runnable task) {
        for (int i = 0; i < mTasks.size(); /* empty */) {
            TaskInfo info = mTasks.get(i);
            if (info.task.equals(task)) {
                if (mLogEnabled) {
                    Timber.tag(TAG).d("[%s] task removed: %s", mOwnerTag, info);
                }
                mTasks.remove(i);
            } else {
                i++;
            }
        }
    }

    @MainThread
    private void checkTasks() {
        mCheckCount++; // for test only
        if (mTasks.isEmpty()) {
            if (mLogEnabled) {
                Timber.tag(TAG).d("[%s] Tasks empty, cancel check.", mOwnerTag);
            }
            mMainHandler.removeMessages(MSG_CHECK_TASKS);
            return;
        }

        if (mLogEnabled) {
            Timber.tag(TAG).v("[%s] check tasks, taskCount: %d", mOwnerTag, mTasks.size());
        }
        Iterator<TaskInfo> it = mTasks.iterator();
        ArrayList<Runnable> pendingTasks = new ArrayList<>();
        long nextEventDelay = mCheckInterval;
        while (it.hasNext()) {
            TaskInfo info = it.next();
            if (SystemClock.elapsedRealtime() >= info.triggerAt) {
                if (mLogEnabled) {
                    Timber.tag(TAG).d("[%s] task to execute: %s", mOwnerTag, info);
                }
                pendingTasks.add(info.task);
                if (info.period > 0) {
                    info.triggerAt = SystemClock.elapsedRealtime() + info.period;
                } else {
                    it.remove();
                    info = null; // mark it removed from queue
                }
            }

            if (info != null) {
                long timeout = info.triggerAt - SystemClock.elapsedRealtime();
                if (timeout < nextEventDelay) {
                    nextEventDelay = timeout;
                }
            }
        }

        if (mLogEnabled) {
            Timber.tag(TAG).v("[%s] next check at %s", mOwnerTag,
                    DateTimeUtils.getReadableTimeStamp(System.currentTimeMillis() + nextEventDelay));
        }
        mMainHandler.removeMessages(MSG_CHECK_TASKS);
        mMainHandler.sendEmptyMessageDelayed(MSG_CHECK_TASKS, nextEventDelay);

        if (pendingTasks.size() > 0) {
            mTaskExecutor.postTasks(pendingTasks);
        }
    }

    @MainThread
    private void clearTasks() {
        mTasks.clear();
        mTaskExecutor.clearTasks();
    }

    @MainThread
    private void scheduleCheckTask(long delay) {
        if (delay > mCheckInterval) {
            delay = mCheckInterval;
        }
        mMainHandler.sendEmptyMessageDelayed(MSG_CHECK_TASKS, delay);
    }

    @SuppressLint("HandlerLeak")
    private class MainHandler extends Handler {
        MainHandler() {
            super(Looper.getMainLooper());
        }

        @Override
        public void handleMessage(Message msg) {
            switch (msg.what) {
                case MSG_ADD_TASK: {
                    TaskInfo task = (TaskInfo) msg.obj;
                    int policy = msg.arg1;
                    addTask(task, policy);
                    break;
                }

                case MSG_REMOVE_TASK: {
                    Runnable task = (Runnable) msg.obj;
                    removeTask(task);
                    break;
                }

                case MSG_CHECK_TASKS: {
                    checkTasks();
                    break;
                }

                case MSG_CLEAR_TASKS: {
                    clearTasks();
                    break;
                }
            }
        }
    }
}

class TaskInfo {
    private static AtomicInteger sTaskId = new AtomicInteger(1);

    private int taskId;
    Runnable task;
    long delay;
    long period = -1;
    long triggerAt;

    TaskInfo(@NonNull Runnable task, long delay) {
        this.taskId = sTaskId.getAndIncrement();
        this.task = task;
        this.delay = delay;
        this.triggerAt = SystemClock.elapsedRealtime() + delay;
    }

    TaskInfo(@NonNull Runnable task, long delay, long period) {
        this.taskId = sTaskId.getAndIncrement();
        this.task = task;
        this.delay = delay;
        this.period = period;
        this.triggerAt = SystemClock.elapsedRealtime() + delay;
    }

    @Override
    public String toString() {
        long timestamp = System.currentTimeMillis() - (SystemClock.elapsedRealtime() - triggerAt);
        return "TaskInfo[id=" + taskId + ", delay=" + delay
                + ", triggerAt=" + DateTimeUtils.getReadableTimeStamp(timestamp)
                + ", period=" + period + "]";
    }
}

File: baseLib/src/main/java/me/ycdev/android/lib/common/async/TaskScheduler.kt
package me.ycdev.android.lib.common.async

import android.annotation.SuppressLint
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.os.SystemClock
import androidx.annotation.IntDef
import androidx.annotation.NonNull
import androidx.annotation.VisibleForTesting
import me.ycdev.android.lib.common.annotation.HandlerWork
import me.ycdev.android.lib.common.utils.DateTimeUtils
import me.ycdev.android.lib.common.utils.Preconditions
import timber.log.Timber
import java.util.ArrayList
import java.util.concurrent.atomic.AtomicInteger

class TaskScheduler(@NonNull schedulerLooper: Looper, @NonNull ownerTag: String) {
    private val ownerTag: String = taskSchedulerIdGenerator.incrementAndGet().toString() + "-" + ownerTag
    private var checkInterval = DEFAULT_CHECK_INTERVAL
    private var logEnabled = false

    private val schedulerHandler = SchedulerHandler(schedulerLooper)
    private val tasks = ArrayList<TaskInfo>()

    // for test only
    @VisibleForTesting
    internal var checkCount: Int = 0

    init {
        Preconditions.checkNotNull(schedulerLooper)
        Preconditions.checkNotNull(ownerTag)
    }

    fun setCheckInterval(interval: Long) {
        if (interval < 1000) {
            throw IllegalArgumentException("Interval less than 1 second is not allowed.")
        }
        checkInterval = interval
    }

    fun enableDebugLogs(enable: Boolean) {
        logEnabled = enable
    }

    fun schedule(executor: ITaskExecutor, delayedMs: Long, task: Runnable) {
        schedule(executor, delayedMs, SCHEDULE_POLICY_NO_CHECK, task)
    }

    fun schedule(
        executor: ITaskExecutor,
        delayedMs: Long,
        @SchedulePolicy policy: Int,
        task: Runnable,
    ) {
        checkSchedulePolicy(policy)
        val taskInfo = TaskInfo(executor, task, delayedMs)
        if (logEnabled) {
            Timber.tag(TAG).d(
                "[%s] schedule one-off task: %s, policy: %s",
                ownerTag, taskInfo, schedulePolicyToString(policy)
            )
        }
        scheduleTask(taskInfo, policy)
    }

    fun schedulePeriod(executor: ITaskExecutor, delayedMs: Long, periodMs: Long, task: Runnable) {
        schedulePeriod(executor, delayedMs, periodMs, SCHEDULE_POLICY_NO_CHECK, task)
    }

    fun schedulePeriod(
        executor: ITaskExecutor,
        delayedMs: Long,
        periodMs: Long,
        @SchedulePolicy policy: Int,
        task: Runnable,
    ) {
        checkSchedulePolicy(policy)
        val taskInfo = TaskInfo(executor, task, delayedMs, periodMs)
        if (logEnabled) {
            Timber.tag(TAG).d(
                "[%s] schedule period task: %s, policy: %s",
                ownerTag, taskInfo, schedulePolicyToString(policy)
            )
        }
        scheduleTask(taskInfo, policy)
    }

    private fun scheduleTask(taskInfo: TaskInfo, @SchedulePolicy policy: Int) {
        if (Looper.myLooper() == schedulerHandler.looper) {
            addTask(taskInfo, policy)
        } else {
            schedulerHandler.obtainMessage(MSG_ADD_TASK, policy, 0, taskInfo).sendToTarget()
        }
    }

    fun cancel(task: Runnable) {
        if (logEnabled) {
            Timber.tag(TAG).d("[%s] cancel task: %s", ownerTag, task)
        }
        if (Looper.myLooper() == schedulerHandler.looper) {
            removeTask(task)
        } else {
            schedulerHandler.obtainMessage(MSG_REMOVE_TASK, task).sendToTarget()
        }
    }

    fun clear() {
        if (logEnabled) {
            Timber.tag(TAG).d("[%s] clear tasks", ownerTag)
        }
        if (Looper.myLooper() == schedulerHandler.looper) {
            clearTasks()
        } else {
            schedulerHandler.sendEmptyMessage(MSG_CLEAR_TASKS)
        }
    }

    fun trigger() {
        if (logEnabled) {
            Timber.tag(TAG).d("[%s] trigger checking", ownerTag)
        }
        if (Looper.myLooper() == schedulerHandler.looper) {
            checkTasks()
        } else {
            schedulerHandler.sendEmptyMessage(MSG_CHECK_TASKS)
        }
    }

    @HandlerWork("schedulerHandler")
    private fun addTask(task: TaskInfo, @SchedulePolicy policy: Int) {
        var taskAdded = false
        if (policy == SCHEDULE_POLICY_NO_CHECK) {
            tasks.add(task)
            taskAdded = true
        } else {
            val index = findTaskIndex(task.task)
            if (index == -1) {
                tasks.add(task)
                taskAdded = true
            } else {
                if (logEnabled) {
                    Timber.tag(TAG).d("[%s] duplicate task found when add %s", ownerTag, task)
                }
                if (policy == SCHEDULE_POLICY_REPLACE) {
                    tasks[index] = task
                    taskAdded = true
                } // else: nothing to do for ignore
            }
        }

        if (taskAdded) {
            scheduleCheckTask(task.delay)
            if (logEnabled) {
                Timber.tag(TAG).d(
                    "[%s] addTask: %s, policy: %s",
                    ownerTag, task, schedulePolicyToString(policy)
                )
            }
        }
    }

    @HandlerWork("schedulerHandler")
    private fun findTaskIndex(task: Runnable): Int {
        for (i in tasks.indices) {
            val info = tasks[i]
            if (info.task == task) {
                return i
            }
        }
        return -1
    }

    @HandlerWork("schedulerHandler")
    private fun removeTask(task: Runnable) {
        var i = 0
        while (i < tasks.size) {
            val info = tasks[i]
            if (info.task == task) {
                if (logEnabled) {
                    Timber.tag(TAG).d("[%s] task removed: %s", ownerTag, info)
                }
                tasks.removeAt(i)
            } else {
                i++
            }
        } /* empty */
    }

    @HandlerWork("schedulerHandler")
    private fun checkTasks() {
        checkCount++ // for test only
        if (tasks.isEmpty()) {
            if (logEnabled) {
                Timber.tag(TAG).d("[%s] Tasks empty, cancel check.", ownerTag)
            }
            schedulerHandler.removeMessages(MSG_CHECK_TASKS)
            return
        }

        if (logEnabled) {
            Timber.tag(TAG).v("[%s] check tasks, taskCount: %d", ownerTag, tasks.size)
        }
        val it = tasks.iterator()
        val pendingTasks = ArrayList<TaskInfo>()
        var nextEventDelay = checkInterval
        while (it.hasNext()) {
            var info: TaskInfo? = it.next()
            if (SystemClock.elapsedRealtime() >= info!!.triggerAt) {
                if (logEnabled) {
                    Timber.tag(TAG).d("[%s] task to execute: %s", ownerTag, info)
                }
                pendingTasks.add(info)
                if (info.period > 0) {
                    info.triggerAt = SystemClock.elapsedRealtime() + info.period
                } else {
                    it.remove()
                    info = null // mark it removed from queue
                }
            }

            if (info != null) {
                val timeout = info.triggerAt - SystemClock.elapsedRealtime()
                if (timeout < nextEventDelay) {
                    nextEventDelay = timeout
                }
            }
        }

        if (logEnabled) {
            Timber.tag(TAG).v(
                "[%s] next check at %s", ownerTag,
                DateTimeUtils.getReadableTimeStamp(System.currentTimeMillis() + nextEventDelay)
            )
        }
        schedulerHandler.removeMessages(MSG_CHECK_TASKS)
        schedulerHandler.sendEmptyMessageDelayed(MSG_CHECK_TASKS, nextEventDelay)

        for (info in pendingTasks) {
            info.executor.postTask(info.task)
        }
    }

    @HandlerWork("schedulerHandler")
    private fun clearTasks() {
        tasks.clear()
    }

    @HandlerWork("schedulerHandler")
    private fun scheduleCheckTask(delay: Long) {
        var delayTmp = delay
        if (delayTmp > checkInterval) {
            delayTmp = checkInterval
        }
        schedulerHandler.sendEmptyMessageDelayed(MSG_CHECK_TASKS, delayTmp)
    }

    @SuppressLint("HandlerLeak")
    private inner class SchedulerHandler(@NonNull looper: Looper) : Handler(looper) {
        override fun handleMessage(msg: Message) {
            when (msg.what) {
                MSG_ADD_TASK -> {
                    val task = msg.obj as TaskInfo
                    val policy = msg.arg1
                    addTask(task, policy)
                }

                MSG_REMOVE_TASK -> {
                    val task = msg.obj as Runnable
                    removeTask(task)
                }

                MSG_CHECK_TASKS -> {
                    checkTasks()
                }

                MSG_CLEAR_TASKS -> {
                    clearTasks()
                }
            }
        }
    }

    @Retention(AnnotationRetention.SOURCE)
    @IntDef(SCHEDULE_POLICY_NO_CHECK, SCHEDULE_POLICY_IGNORE, SCHEDULE_POLICY_REPLACE)
    annotation class SchedulePolicy

    companion object {
        private const val TAG = "TaskScheduler"

        const val SCHEDULE_POLICY_NO_CHECK = 1
        const val SCHEDULE_POLICY_IGNORE = 2
        const val SCHEDULE_POLICY_REPLACE = 3

        private const val MSG_ADD_TASK = 1
        private const val MSG_REMOVE_TASK = 2
        private const val MSG_CHECK_TASKS = 3
        private const val MSG_CLEAR_TASKS = 4

        @VisibleForTesting
        internal const val DEFAULT_CHECK_INTERVAL: Long = 10_000 // 10 seconds

        private val taskSchedulerIdGenerator = AtomicInteger(0)

        private fun schedulePolicyToString(@SchedulePolicy policy: Int): String {
            return when (policy) {
                SCHEDULE_POLICY_NO_CHECK -> "NO_CHECK"
                SCHEDULE_POLICY_IGNORE -> "IGNORE"
                SCHEDULE_POLICY_REPLACE -> "REPLACE"
                else -> throw RuntimeException("Unknown policy: $policy")
            }
        }

        private fun checkSchedulePolicy(@SchedulePolicy policy: Int) {
            when (policy) {
                SCHEDULE_POLICY_NO_CHECK,
                SCHEDULE_POLICY_IGNORE,
                SCHEDULE_POLICY_REPLACE -> return
                else -> throw RuntimeException("Unknown policy: $policy")
            }
        }
    }
}

